### PR TITLE
unittest tests should test actual kustomizations.

### PR DESCRIPTION
**Pull request by jlewi**
* generate_legacy_tests.py is a onetime script intended to generate
  kustomization.yaml files that are nearely identical to the ones that
  get generated by kfctl. These kustomization.yaml files can be used to
  generate the golden/expected output. We can use these to test that
  we actually produce the expected output.

* This will make it easier to refactor our kustomize packages and verify
  that we haven't changed the existing kustomizations.

* Prior to this PR; generate_tests.py looked for every kustomization.yaml
  file in the manifests repo and generated a test for that kustomization.yaml
  file. That doesn't really do what we want because many of those kustomization.yaml files are trivial so we don't end up testing any transformations.

  * Right now the transformations we want to test are the ones applied by the
    kustomization.yaml files created by kfctl

  * Thats why we check in kustomization.yaml files corresponding to the
    kustomization.yaml files generated by kfctl.

  * Once we start using kustomize to compose applications into packages
    we will test that those stacks generate the expected output.

* Related to #1014

* No longer try to infer what tests were modified; instead just rerun
  generate all. Now that we are testing derived kustomizations its no
  longer easy to infer based on which files changed which tests to update.

### CODE REVIEW

File: hack/generate_legacy_tests.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    p_files = g.get("envs", [])
+
+    if "env" in g:
+      p_files.append(g["env"])
+
+    generators[g["name"]] = [os.path.join(kustomize_dir, f) for f in p_files]
+
+  return generators
+
+class GenerateLegacyTests:
```
[the remark this answers is not in the document]
Comment (jlewi) [reply hack/generate_legacy_tests.py:108]: I should probably rename this to generate_legacy_kustomizations.py.
The tests are actually generated by generate_tests.py. In that directory you list the directories to search for kustomization.yaml to generate tests for.
https://github.com/kubeflow/manifests/blob/9cbb4d721d7b9e62da155956e3f2eae2a25e8b71/hack/generate_tests.py#L11

So if we have kustomizations that are worth testing we just include in the search directory.

The purpose of generate_legacy_tests.py is to account for the fact that right now we don't check our kustomization.yaml files into kubeflow/manifests and instead generate them via kfctl magic (#774).

So the way our tests are currently structured we generate a kustomization.yaml for kustomization.yaml we find in kubeflow/manifests. This kustomizations end up being trivial (i.e. not combining overlays or doing var substitutions). So we don't really end up testing that when all of the kustomizations are applied that we generate the expected YAML which is what we want to test.

File: hack/generate_legacy_tests.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+import fire
+import logging
+import os
+import shutil
+import yaml
+
+# Which apps to skip. These are apps which had some edge case that it wasn't
+# worth dealing with.  It would just make more sense to generate
+# the tests manually
+APPS_TO_SKIP = ["mysql", "minio"]
```
[the remark this answers is not in the document]
Comment (jlewi) [reply hack/generate_legacy_tests.py]: Added more explanation. PTAL

**Comment by jlewi** [comment]
@krishnadurai ping?

File: hack/generate_legacy_tests.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (krishnadurai) [reply hack/generate_legacy_tests.py]: Thanks

File: hack/generate_legacy_tests.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (krishnadurai) [reply hack/generate_legacy_tests.py]: Thanks for the explanation.
If I'm understanding correctly:
1. Until the time stacks are implemented, we would be following this current approach of testing.
2. If we looking to generate a comprehensive test set we should look to use all supported kfdefs.
3. We may have to manually check-in kustomizations which are not referred to by any kfdef in the path `tests/legacy_kustomizations/`.

It seems the kustomizations for kfctl_k8s_istio.yaml or kfctl_istio_dex.yaml are yet to be generated. We should generate them with this PR.

It might be fine to overlook kustomizations affected by point 3.

[the remark this answers is not in the document]
Comment (krishnadurai) [reply hack/generate_legacy_tests.py]: I understand that the conflicting bit here seems to be the kustomization.yaml file being generated by kfctl. We can call it the unit which we are testing here. The expectation from a developer should be to check in a kustomization.yaml for each 'unit' module.

Some documentation on the README.md file can help.

**Comment by jlewi** [comment]
@krishnadurai this PR strictly increases test coverage. So I think any improvements to increase coverage e.g. cover additional KFDef's can be done in follow on PRs.

Right now we don't have any coverage to ensure that any refactoring of the kustomize packages doesn't break existing KFDefs. Even though this PR is only testing the KFDef from GCP this should give us relatively high coverage because for individual applications e.g. jupyterweb app there is little difference in how they are configured for different KFDefs.

File: hack/generate_legacy_tests.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jlewi) [reply hack/generate_legacy_tests.py]: Per comment below I don't think we should generate additional coverage in this PR. This PR is strictly increasing coverage. The priority is to increase converage to unblock the refactoring of the kustomize manifests for the V3 kustomize patterns.

So I think we should merge this PR and then any additional test coverage (e.g. for additional KFDefs) can happen in parallel.

> The earlier version of the unit test mechanism used to ensure that if make generate is run, the added kustomization's test files are automatically added to the test folder. This would ensure coverage or inform a developer to add tests.

This mechanism still exists. `make generate` will automatically generate tests for all paths listed in https://github.com/kubeflow/manifests/blob/9cbb4d721d7b9e62da155956e3f2eae2a25e8b71/hack/generate_tests.py#L11.

Right now the only search path is "legacy_kustomizations" because those are the only kustomization.yaml's that are worth testing. All the other kustomization.yaml files are trivial.

Think of it this way. imagine we create "kubeflow/manifests/examples" and in that directory we start definining kustomizations that we provide as examples to customers. So imagine we have a kustomization for katib that does something complicated; e.g use 5 different overlays, 6 different patches and we want to verify that it all generates the expected YAML.

At that point we would just add "kubeflow/manifests/examples" to search_dirs and then we would start autogenerating tests for those kustomization.yamls.

This should become more clear in subsequent PRs as we will start checking in complicated kustomization.yamls and creating tests for them by adding them to generate_tests.py

**Comment by jlewi** [comment]
Thanks @krishnadurai 
/approve
